File: Data Structures & Algorithms/climbing-stairs/submission-7.kt
class Solution {
    fun climbStairs(n: Int): Int {
        val cache = IntArray(n + 1) { -1 }

        fun recursive(k: Int): Int {
        if(k > n) return 0
        if(k == n) return 1
        if(cache[k] != -1) return cache[k]

        cache[k] = recursive(k + 1) + recursive(k + 2)
        return cache[k]
    }
        return recursive(0)
    }
}

File: Data Structures & Algorithms/clone-graph/submission-7.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
// Recursive DFS
class Solution {
   public:
    Node* cloneGraph(Node* node) {
        map<Node*, Node*> hashMap;
        return dfs(node, hashMap);
    }

   private:
    Node* dfs(Node* node, map<Node*, Node*>& hashMap) {
        if (node == nullptr) {
            return nullptr;
        }
        if (hashMap.count(node)) {
            return hashMap[node];
        }

        Node* clone = new Node(node->val);
        hashMap[node] = clone;

        for (Node* nei : node->neighbors) {
            clone->neighbors.push_back(dfs(nei, hashMap));
        }

        return clone;
    }
};
